gto/git_utils.py

```diff
@@ -116,6 +116,28 @@ def wrapped_f(*args, **kwargs):
     return wrap
 
 
+def push_on_auto_push(f: Callable):
```

Comment:
I see that we use different approaches to push Git references for `assign` (and others) and `annotate` (and others). Could we maybe unite them? E.g. the API method should return Event(s), and wrapper should get Git references from those events, and push? E.g. the same wrapper would invoke `git push origin abc1234` after annotate, and `git push origin nn@v0.0.1` after `register`. WDYT?

Reply:
This approach would also do `push` only after the whole gto command succeeds. E.g. if you run `assign` and for some reason, the version was registered, but stage wasn't assigned, it won't run `git push` and let you fix things locally before impacting the remote.

Reply:
This may require some other changes though, since the `_return_event` that's invoked in the end of each API call can't currently handle the Git tags we remove. Only the created ones, and I believe if in `assign` we create 2 tags (registration + assignment), it will return the second only.

Reply:
So, this would require changing the interface of the commands in `api.py`, correct?

Reply:
Yes, I believe, what is `return`ed may require a change.

Reply:
> check the tags before the command is executed, then after the command is executed, and push only the difference.

Sounds as a good idea, but don't like that if the repo will change in the meantime (while the command runs), that will result in a wrong diff.

Reply:
> Then I would suggest to leave out the decision for another round

Do you suggest handling this in a separate issue then? I guess it could be in the same place as the other refactoring things (renaming functions, possibly using something but the functional approach, etc).

Reply:
Yes, that's what I am suggesting. It's anyway quite a mess, so let's get the thing to work, and then refine the code. This is what we have done until now, and it has the positive side that now we see the full picture. At the beginning it was unclear to me where we would have ended up.

Reply:
> Sounds as a good idea, but don't like that if the repo will change in the meantime (while the command runs), that will result in a wrong diff.

Oh god, if that happens we are screwed anyway, aren't we? (e.g. registering model v0.0.1, but in the meantime that tag is created by another process) :D Ideally one should lock the repo and unlock the repo while the command is running

Reply:
Maybe locking is a good option. Still, unclear to me which approach is better :)

Reply:
Let's discuss it further, I am also not 100% sure what is best. Need to think deeper about it, will come to this point, it's an important design decision, thanks for rising it :)

Just, let's do it in a separate PR :)